src/ufuncs.c

```diff
@@ -1844,6 +1850,78 @@ static void polygonize_func(char** args, npy_intp* dimensions, npy_intp* steps,
 }
 static PyUFuncGenericFunction polygonize_funcs[1] = {&polygonize_func};
 
+static char polygonize_full_dtypes[5] = {NPY_OBJECT, NPY_OBJECT, NPY_OBJECT, NPY_OBJECT,
+                                         NPY_OBJECT};
+static void polygonize_full_func(char** args, npy_intp* dimensions, npy_intp* steps,
+                                 void* data) {
+  GEOSGeometry* geom = NULL;
+  GEOSGeometry* geom_copy = NULL;
+  unsigned int n_geoms;
+
+  GEOSGeometry* collection = NULL;
+  GEOSGeometry* cuts = NULL;
+  GEOSGeometry* dangles = NULL;
+  GEOSGeometry* invalidRings = NULL;
+
+  GEOS_INIT;
+
+  GEOSGeometry** geoms = malloc(sizeof(void*) * dimensions[1]);
+  if (geoms == NULL) {
+    errstate = PGERR_NO_MALLOC;
+    goto finish;
+  }
+
+  SINGLE_COREDIM_LOOP_OUTER_NOUT4 {
+    n_geoms = 0;
+    SINGLE_COREDIM_LOOP_INNER {
+      if (!get_geom(*(GeometryObject**)cp1, &geom)) {
+        errstate = PGERR_NOT_A_GEOMETRY;
+        goto finish;
+      }
+      if (geom == NULL) {
+        continue;
+      }
+      // need to copy the input geometries, because the Collection takes ownership
+      geom_copy = GEOSGeom_clone_r(ctx, geom);
+      if (geom_copy == NULL) {
+        errstate = PGERR_GEOS_EXCEPTION;
```

Comment:
There may be an edge case here if `GEOSGeom_clone_r` fails partway through.  There will be some geometry objects in `geoms` (up to `n_geoms`) that need to be destroyed before freeing the array.

To be clear, this should only be done if things fail at this step.  As soon as ownership of the cloned geometries transfers to the new GeometryCollection, things are OK (they are properly deleted on failure of creation of GeometryCollection)

Reply:
I added some code for this, but I don't think this will actually be covered? (I am not sure how a clone can fail, except something like out of memory)

Reply:
Agreed - not sure how to make it fail - but seems good from a defensive programming standpoint even if we can't get coverage for this.

It would be possible to proof this with a manual test: just fail if `i == 2` in order to verify that the implementaion is correct, then take that back out (it looks correct).

Reply:
Making the check above `if ((geom_copy == NULL) | (n_geoms == 2)) {` gives a "correct" GEOSException in python, so the destroy code should have run fine, I suppose.